helix-term/src/keymap.rs

```diff
@@ -198,13 +198,40 @@ impl<'de> serde::de::Visitor<'de> for KeyTrieVisitor {
     where
         M: serde::de::MapAccess<'de>,
     {
+        let mut label = "";
+        let mut command = None;
         let mut mapping = HashMap::new();
         let mut order = Vec::new();
-        while let Some((key, value)) = map.next_entry::<KeyEvent, KeyTrie>()? {
-            mapping.insert(key, value);
-            order.push(key);
+
+        while let Some(key) = map.next_key::<&str>()? {
+            match key {
+                "label" => label = map.next_value::<&str>()?,
+                "command" => command = Some(map.next_value::<MappableCommand>()?),
+                _ => {
+                    let key_event = key.parse::<KeyEvent>().map_err(serde::de::Error::custom)?;
+                    let key_trie = map.next_value::<KeyTrie>()?;
+                    mapping.insert(key_event, key_trie);
+                    order.push(key_event);
+                }
+            }
+        }
+
+        match command {
+            None => Ok(KeyTrie::Node(KeyTrieNode::new(label, mapping, order))),
+            Some(cmd) => {
+                let status = (cmd, label.is_empty());
+                if let (MappableCommand::Typable { name, args, .. }, false) = status {
+                    Ok(MappableCommand::Typable {
+                        name,
+                        args,
+                        doc: label.to_string(),
+                    })
+                    .map(KeyTrie::Leaf)
+                } else {
+                    Ok(KeyTrie::Leaf(status.0))
+                }
+            }
```

Comment:
```suggestion
            Some(MappableCommand::Typable { name, args, .. }) if !label.is_empty() => {
                Ok(KeyTrie::Leaf(MappableCommand::Typable {
                    name,
                    args,
                    doc: label.to_string(),
                }))
            }
            Some(command) => Ok(KeyTrie::Leaf(command)),
```

Reply:
Thanks! I kept trying to simplify this and this not making much progress. I did not realize Rust had pattern match guards, and didn't even think to check. Too much Elm, not enough Elixir & Erlang lately, I guess.